### PR TITLE
percy 0.0.5 with patch support

### PR DESCRIPTION
**Pull request by cbouss**
Changes:
- Update percy to 0.0.5 (few bug fixes, recipe patch support)
- When not run from aggregate, use a conda_build_config.yaml copy as configuration to identify variants
- Run linter on all rendered variants
- Add --fix option to linter
- Implement fix for missing_wheel, pip_install_args, avoid_noarch, missing_pip_check, missing_test_requirement_pip, missing_python, remove_python_pinning

I don't have code testing the fix methods at this point.
As a test you can run generate a recipe with grayskull and call conda-lint with --fix.

### CODE REVIEW

**Comment by schuylermartin45** [comment]
Added Mark for visibility

File: anaconda_linter/lint/__init__.py [hunk 1]
```diff
@@ -676,43 +683,64 @@ def lint_file(
...
+                    recipe_path=meta_yaml,
+                    subdir=arch_name,
+                    variant_config_files=variant_config_files,
+                    exclusive_config_files=exclusive_config_files,
+                )
             else:
-                # for when no cbc is provided
-                vid = "dummy"
-                variant = self.config[arch_name]
-            recipe = _recipe.Recipe.from_file(
```
Comment (schuylermartin45) [inline anaconda_linter/lint/__init__.py:695]: I'm not sure if this is a rendering issue on GitHub, but these lines (695 to 700) don't look indented to the same level as the rest of the else (692 to 694). Just want to sanity check that before it causes a runtime error in this case.

File: anaconda_linter/lint/__init__.py [hunk 2]
```diff
@@ -676,43 +683,64 @@ def lint_file(
...
-                recipe_fname=str(meta_yaml),
-                variant_id=vid,
-                variant=variant,
-                renderer=RendererType.RUAMEL,
-            )
-        except _recipe.RecipeError as exc:
+                logging.debug(
+                    "No cbc in current path. Loading copy of aggregate cbc embedded in linter."
+                )
+                logging.debug("Please run from your aggregate dir for better results.")
```
Comment (schuylermartin45) [inline anaconda_linter/lint/__init__.py:702]: Not sure how I feel about this attempted recovery mechanism. I suppose it's safe if we catch the exception, I just don't know how much help we want to give the user before they are flat out "doing it wrong".

File: anaconda_linter/run.py [hunk 3]
```diff
@@ -82,6 +82,13 @@ def check_path(value):
         action="store_true",
         help="Enable verbose output. This displays all of the checks that the linter is running.",
     )
+    # this is here because we have a different default than build
+    parser.add_argument(
+        "-f",
+        "--fix",
+        action="store_true",
+        help="Attempt to fix issues.",
```
Comment (schuylermartin45) [inline anaconda_linter/run.py:90]: I feel like we should have a little bit more information here. Thinking like a new user: Are warnings patched or just errors? Are all errors currently supported?

File: anaconda_linter/lint/check_build_help.py [hunk 4]
```diff
@@ -380,10 +388,23 @@ def check_deps(self, deps):
...
                         for num, line in enumerate(buildsh):
                             if not self._check_line(line):
-                                self.message(fname=build_file, line=num, output=output)
+                                self.message(
+                                    fname=build_file, line=num, output=output, data=self.recipe
+                                )
                 except FileNotFoundError:
                     pass
 
+    def fix(self, _message, recipe):
```
Comment (schuylermartin45) [inline anaconda_linter/lint/check_build_help.py:397]: I was hoping `fix` support would be added with a member class function.

Could/should we enforce the inclusion of this method with the abstract base class decorator? Are there other member functions we should strictly enforce the inclusion of that way?

File: anaconda_linter/lint/__init__.py (hunk 1 above)
Comment (cbouss) [reply anaconda_linter/lint/__init__.py:695]: That is in removed code. Nonetheless the indentation was correct.

A variant is a minimum set of values that result in a different rendition of a recipe. For example, supporting openssl 1.1.1 and 3 results in doubling the number of variants for recipes using openssl.

Until now, the linter was not run on all renditions. It was only doing one per subdir. With this change, all renditions are linted. This is important because based on selectors, we may otherwise miss on some issues.

To compute the variants, conda_build_config.yaml is required. Ideally, same as with conda-build, conda-lint is called from aggregate, and the conda_build_config.yaml from aggregate is used. To still provide okay results when that's not the case, I have added a copy of conda_build_config.yaml to the linter, to be used when not running from aggregate.

Comment (schuylermartin45) [reply anaconda_linter/lint/__init__.py:695]: You can tell it's late on a Friday, didn't even notice it was in a removal. My eyes played tricks on me and I thought the `except` was on the same level as the `else`. Thanks for the sanity check, I clearly needed it :P

**Review by varlackc — APPROVED** [review]
LGTM

File: anaconda_linter/run.py (hunk 3 above)
Comment (schuylermartin45) [reply anaconda_linter/run.py:90]: I like that idea a lot.

Probably a future issue, but we could also have a dry-run mode that shows a diff of changes before we commit then with `-f`? I guess a user could also revert them in `git`, so I'm not sure how useful that would be.

File: anaconda_linter/lint/check_build_help.py (hunk 4 above)
Comment (schuylermartin45) [reply anaconda_linter/lint/check_build_help.py:397]: That's fair. Maybe a problem for another time, but if any of those functions are 100% required, I think they should be enforced.

Maybe in the future we get to the point of 100% fixability and then we can enforce inclusion of the overridden methods to ensure developers don't forget to implement it.

File: setup.py [hunk 5]
```diff
@@ -20,7 +20,6 @@ def main():
         "networkx",
         "requests",
         "ruamel.yaml",
-        "percy==0.0.3",
```
Comment (marcoesters) [inline setup.py:23]: Why is this removed?

File: anaconda_linter/lint/check_build_help.py [hunk 6]
```diff
@@ -736,15 +838,27 @@ class remove_python_pinning(LintCheck):
...
+        for package in recipe.packages.values():
+            if recipe.get(f"{package.path_prefix}build/noarch", "") == "":
+                for section in ["host", "run"]:
+                    for dep in package.get(section):
+                        if dep.pkg == "python" and dep.constraint != "":
+                            self.message(section=dep.path, data=(recipe, package, dep))
+
+    def fix(self, _message, _data):
+        (recipe, package, dep) = _data
+        s = dep.split(">=")
```
Comment (marcoesters) [inline anaconda_linter/lint/check_build_help.py:850]: While rare, `<=` or `==` or `!=` are also potential operators we may encounter. That may be beyond the scope of this PR though

Comment (cbouss) [reply anaconda_linter/lint/check_build_help.py:861]: addressed in new commit, ty

File: anaconda_linter/run.py (hunk 3 above)
Comment (cbouss) [reply anaconda_linter/run.py:90]: That's what I used for testing. With the file opened on vscode, cmd+z is also sufficient to revert changes.

**Review by marcoesters — APPROVED** [review]
